File: BFS_DFS/46_Permutations/main.py
def permute(self, nums):
    """
    :type nums: List[int]
    :rtype: List[List[int]]
    """
    res = []
    if not nums: return []
    self.dfs(nums, [], res) 
    return res

def dfs(self, nums, path, res):
    if not nums:
        res.append(path)
        return
    
    for i in range(len(nums)):
        self.dfs(nums[:i]+nums[i+1:], path+[nums[i]], res)
def dfs(self, nums, path, res):
    if not nums:
        res.append(path[:])
        return
    
    for i in range(len(nums)):
        path.append(nums[i])
        self.dfs(nums[:i]+nums[i+1:], path, res)
        path.pop()

File: BFS_DFS/46_Permutations/main.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> perms;
        permute(nums, 0, perms);
        return perms;
    }
private:
    void permute(vector<int> nums, int i, vector<vector<int>>& perms) {
        if (i == nums.size()) {
            perms.push_back(nums);
        } else {
            for (int j = i; j < nums.size(); j++) {
                swap(nums[i], nums[j]);
                permute(nums, i + 1, perms);
            }
        }
    }
};